outlines/generate/generator.py

```diff
@@ -0,0 +1,359 @@
+    token_generator: Callable,
+    fsm: "FSM",
+    init_state: Tuple,
+    fsm_states: List[FSMState],
+    rng: torch.Generator,
+) -> Iterator[Sequences]:
+    """Generates sequences of tokens.
+
+    Parameters
+    ----------
+    token_generator
+        A callable that generate a new token given the current generation state
+        and logits biases.
+    fsm
+        The finite-state machine that drives the text generation.
+    init_state
+        The initial generation state for the batches.
+    fsm_states
+        The initial states of the finite-state machine for each sequence in the batch.
+
+    Yields
+    ------
+    A new sequence.
+
+    """
+    token_ids, attention_masks, kv_cache = init_state
+    while True:
+        logits_masks = get_next_instruction(fsm, fsm_states)
+
+        next_tokens, next_token_ids, kv_cache, logits = token_generator(
+            token_ids,
+            attention_masks,
+            kv_cache,
+            rng=rng,
+            logits_masks=logits_masks,
+        )
+
+        fsm_states = get_next_fsm_states(fsm, fsm_states, next_token_ids)
+        is_finished = is_generation_finished(fsm, fsm_states)
+
+        token_ids = update_token_ids(token_ids, next_token_ids)
+        attention_masks = expand_attention_masks(attention_masks)
+
+        if is_finished:
+            yield Sequences(next_tokens, token_ids, kv_cache, logits)
+            return
+
+        yield Sequences(next_tokens, token_ids, kv_cache, logits)
+
+
+def token_generator(model, sampler: "Sampler") -> Callable:
+    """Generate one token at a time.
+
+    This process is designed to be steered by another supervising
+    process that supplies the current sequence and the indices
+    of the tokens to mask before sampling.
+
+    Parameters
+    ----------
+    model
+        A model that takes a sequence of tokens as an input and
+        returns a probability distribution over the next tokens.
+    sampler
+        A function that samples tokens from a probability
+        distribution over the next tokens.
+
+    Returns
+    -------
+    A tuple that contains a tensor with the sampled tokens, a tensor with
+    the K-V cache for the sequence and the tensor that contains the next-token
+    logits that were returned by the model.
+
+    """
+
+    def generate(
+        token_ids,
+        attention_masks,
+        kv_cache,
+        logits_masks,
+        rng: torch.Generator,
+    ):
+        try:
+            logits, new_kv_cache = model(token_ids, attention_masks, kv_cache)
+        except IndexError:  # Exceeding the context length
+            raise IndexError(
+                "The input length exceeds the context length of the model."
+            )
+
+        biased_logits = bias_logits(logits, logits_masks)
+        next_token_ids = sampler(biased_logits, 1, rng)
+        next_tokens = model.tokenizer.decode(next_token_ids)
+
+        return next_tokens, next_token_ids, new_kv_cache, biased_logits
```

Comment:
I think `decode` should happen at a higher level unless we have a very good reason not to.